### PR TITLE
Encode CrateNum using the StableCrateId for incr. comp.

### PR DESCRIPTION
**Pull request by cjgillot**

### CODE REVIEW

File: compiler/rustc_middle/src/ty/query/on_disk_cache.rs [hunk 1]
```diff
@@ -1077,6 +1038,23 @@ where
...
+{
+    fn encode(&self, s: &mut CacheEncoder<'a, 'tcx, E>) -> Result<(), E::Error> {
+        let def_path_hash = s.tcx.def_path_hash(self.as_def_id());
+        // Store additional information when we encode a foreign `DefId`,
+        // so that we can map its `DefPathHash` back to a `DefId` in the next
+        // compilation session.
+        if *self != LOCAL_CRATE {
+            s.latest_foreign_def_path_hashes
+                .insert(def_path_hash, RawDefId { krate: self.as_u32(), index: 0 });
+        }
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/query/on_disk_cache.rs:1053]: Is this necessary?

File: compiler/rustc_middle/src/ty/context.rs [hunk 2]
```diff
@@ -1290,6 +1291,15 @@ impl<'tcx> TyCtxt<'tcx> {
...
     }
 
+    #[inline]
+    pub fn stable_crate_id(self, cnum: CrateNum) -> StableCrateId {
+        if cnum == LOCAL_CRATE {
+            self.sess.local_stable_crate_id()
+        } else {
+            self.cstore.stable_crate_id_untracked(cnum)
+        }
+    }
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/context.rs:1301]: This could be used in many places. I would guess it happens like 20-30 times.

**Comment by bjorn3** [comment]
Mixed small improvements and regressions.

**Comment by michaelwoerister** [comment]
It's on my todo list :)

**Comment by michaelwoerister** [comment]
Looks good to me - thanks, @cjgillot! 

The performance changes are in the noise and the code is simpler than before.

@bors r+

**Pull request merged**